src/vmemcache_heap.c

```diff
@@ -35,128 +35,157 @@
+#define HEADER(ptr) ((struct header *)((uintptr_t)(ptr) - HEADER_SIZE))
+
+/* pointer to the footer */
+#define FOOTER(ptr, size) ((struct footer *)((uintptr_t)(ptr) + (size)))
+
+/* size from the footer */
+#define SIZE_FROM_FOOTER(ptr) (*(uint64_t *)((uintptr_t)ptr))
+
+static void
+vmcache_heap_merge(struct heap *heap, struct extent *ext,
+			struct heap_entry *he);
+
 /*
- * vmcache_heap_write_headers -- write headers of the heap extent
+ * vmcache_extent_get_prev -- get the pointer to the previous extent
  */
-static inline void
-vmcache_heap_write_headers(struct heap_entry *he, int is_allocated)
+void *
+vmcache_extent_get_prev(void *ptr)
 {
-	size_t header_value = HEADER_VALUE(he->size, is_allocated);
-
-	/* save the header */
-	HEADER(he->ptr) = header_value;
-
-	/* save the footer */
-	FOOTER(he->ptr, he->size) = header_value;
+	ASSERTne(ptr, NULL);
+	return HEADER(ptr)->prev;
 }
 
 /*
- * vmcache_heap_verify_headers -- verify headers of the heap extent
+ * vmcache_extent_get_size -- get size of the extent
  */
-static inline void
-vmcache_heap_verify_headers(struct heap_entry *he, int is_allocated)
+size_t
+vmcache_extent_get_size(void *ptr)
 {
-	size_t header_value = HEADER_VALUE(he->size, is_allocated);
-
-	/* verify the header */
-	ASSERTeq(HEADER(he->ptr), header_value);
-
-	/* verify the footer */
-	ASSERTeq(FOOTER(he->ptr, he->size), header_value);
+	ASSERTne(ptr, NULL);
+	return HEADER(ptr)->size_flags & MASK_FLAGS;
 }
 
 /*
- * vmcache_heap_insert -- mark the heap entry as free and insert it to the heap
+ * vmcache_insert_heap_entry -- insert the 'he' entry into the list of extents
  */
 static inline int
-vmcache_heap_insert(struct heap *heap, struct heap_entry *he)
+vmcache_insert_heap_entry(struct heap *heap, struct heap_entry *he,
+				void **last_extent, int is_allocated)
 {
-	vmcache_heap_write_headers(he, IS_FREE);
+	struct header *header = he->ptr;
+	struct footer *footer = VOID_PTR(he->ptr, he->size - FOOTER_SIZE);
 
-	if (VEC_PUSH_BACK(&heap->entries, *he)) {
-		ERR("!cannot grow the heap's vector");
-		return -1;
+	/* pointer and size of a new extent */
+	void *new_extent = VOID_PTR(he->ptr, HEADER_SIZE);
+	size_t size_flags =
+		SIZE_FLAGS((he->size - HFER_SIZE), is_allocated);
+
+	/* save the header */
+	header->prev = *last_extent;
+	header->next = NULL;
+	header->size_flags = size_flags;
+
+	/* save the footer */
+	footer->size_flags = size_flags;
+
+	if (*last_extent) {
+		struct header *last_header = HEADER(*last_extent);
+		ASSERTeq(last_header->next, NULL);
+		last_header->next = new_extent;
+	}
+
+	*last_extent = new_extent;
+
+	if (!is_allocated) {
+		__sync_fetch_and_add(&heap->entries, 1);
 	}
 
 	return 0;
 }
 
 /*
- * vmcache_heap_get -- get the free heap entry from the heap
+ * vmcache_get_heap_entry -- get the free entry from the heap
  */
 static inline int
-vmcache_heap_get(struct heap *heap, struct heap_entry *he)
+vmcache_get_heap_entry(struct heap *heap, struct heap_entry *he)
```

Comment:
this isn't 'get', it's a 'pop' or 'remove' operation.

Reply:
Done